system/system_windows.go

```diff
@@ -72,67 +73,120 @@ func removeUninstallEntry(name string) error {
-	if err != nil {
-		return ""
+func AppAssociation(publisher, installSource, programName, extension string) (string, string) {
+
+	var productroots = []string{
+		`SOFTWARE\Microsoft\Windows\CurrentVersion\Uninstall\`,
+		`SOFTWARE\WOW6432Node\Microsoft\Windows\CurrentVersion\Uninstall\`,
 	}
-	reg, _ := k.ReadSubKeyNames(-1)
-	defer k.Close()
-	for _, v := range reg {
-		q, err := registry.OpenKey(registry.LOCAL_MACHINE, fmt.Sprintf("%s%s", productroot, v), registry.ALL_ACCESS)
-		defer q.Close()
+	for _, productroot := range productroots {
+		k, err := registry.OpenKey(registry.LOCAL_MACHINE, productroot, registry.ENUMERATE_SUB_KEYS)
 		if err != nil {
-			continue
+			return "", ""
 		}
-		switch extension {
-		case "msi":
-			a, _, err := q.GetStringValue("InstallSource")
+		reg, _ := k.ReadSubKeyNames(-1)
+		defer k.Close()
+		for _, v := range reg {
+			productReg := fmt.Sprintf("%s%s", productroot, v)
+			q, err := registry.OpenKey(registry.LOCAL_MACHINE, productReg, registry.ALL_ACCESS)
+			defer q.Close()
 			if err != nil {
-				// InstallSource not found, move on to next entry
 				continue
 			}
-			if strings.Contains(a, installSource) {
-				un, _, err := q.GetStringValue("UninstallString")
-				if err != nil {
-					// UninstallString not found, move on to next entry
-					continue
-				}
-				return un
-			}
-		default:
-			for _, v := range publisherNameReg {
-				re := regexp.MustCompile("(?i)" + regex[v])
-				publisher = re.ReplaceAllString(publisher, "")
-			}
-			for _, v := range programNameReg {
-				re := regexp.MustCompile("(?i)" + regex[v])
-				programName = re.ReplaceAllString(programName, "")
-			}
-			publisherReg, _, err := q.GetStringValue("Publisher")
+			displayName, _, err := q.GetStringValue("DisplayName")
 			if err != nil {
 				continue
 			}
-			// Remove display name whitespace to conform to package name
-			if strings.ToLower(publisherReg) == strings.ToLower(publisher) {
-				displayName, _, _ := q.GetStringValue("DisplayName")
-				if strings.Contains(strings.ToLower(programName), strings.ToLower(strings.ReplaceAll(displayName, " ", ""))) {
-					// Check if the value exists, move on if it doesn't
-					un, _, err := q.GetStringValue("QuietUninstallString")
+
+			switch extension {
+			case ".msi":
+				a, _, err := q.GetStringValue("InstallSource")
+				if err != nil {
+					// InstallSource not found, move on to next entry
+					continue
+				}
+				iS := strings.Split(installSource, "@")
+				if strings.Contains(a, iS[0]) {
+					name, _, err := q.GetStringValue("DisplayName")
 					if err != nil {
-						un, _, err = q.GetStringValue("UninstallString")
-						if err != nil {
-							// UninstallString not found, move on to next entry
-							continue
-						}
+						// UninstallString not found, move on to next entry
+						continue
 					}
-					return un
+					return name, productReg
+				}
+			default:
+				for _, v := range publisherNameReg {
+					re := regexp.MustCompile("(?i)" + regex[v])
```

Comment:
Why aren't the regexes precompiled? And why are they declared in system.go instead of system_windows.go?

Reply:
Note, not asking to change in this PR, but if its something that can be fixed, please leave a TODO for later.